File: frontend/src/services/authService.js
/**
 * Authentication Service
 * Handles all API calls related to auth
 */
import axios from 'axios';

const API_URL = 'http://localhost:8080/api/auth';

/**
 * Send verification code to email for registration
 * @param {string} email - User's email address
 * @returns {Promise} Response from the API
 */
export const sendVerificationCode = async (email) => {
  try {
    const response = await axios.post(`${API_URL}/send-verification-code`, { email: email.trim() });
    return response.data;
  } catch (error) {
    throw error.response?.data || error.message || 'Error sending verification code';
  }
};

/**
 * Verify the code sent to user's email
 * @param {string} email - User's email address
 * @param {string} token - Verification code/token
 * @returns {Promise} Response from the API with token and user ID
 */
export const verifyCode = async (email, token) => {
  try {
    const response = await axios.post(`${API_URL}/verify-verification-code`, { 
      email: email.trim(), 
      token: token.trim() 
    });
    return response.data;
  } catch (error) {
    throw error.response?.data || error.message || 'Error verifying code';
  }
};

/**
 * Update user with password and role
 * @param {string} email - User's email address
 * @param {string} password - User's password
 * @param {string} role - User's role (user, owner, etc.)
 * @returns {Promise} Response from the API
 */
export const updateUser = async (email, password, role = 'user') => {
  try {
    const response = await axios.post(`${API_URL}/update-user`, { 
      email: email.trim(), 
      password: password,
      role: role 
    });
    return response.data;
  } catch (error) {
    throw error.response?.data || error.message || 'Error setting password';
  }
};

/**
 * Login with email and password
 * @param {string} email - User's email address
 * @param {string} password - User's password
 * @param {string} role - User's role
 * @returns {Promise} Response from the API with tokens and user info
 */
export const login = async (email, password, role) => {
  try {
    const response = await axios.post(
        "http://localhost:8080/api/auth/login",
        {
          email: email.trim(),
          password: password.trim(),
          role: role
        }
    );
    return response.data;
  } catch (error) {
    throw error.response?.data || error.message || 'Error logging in';
  }
};
export const checkPassword = async (email, password, role) => {
  try {
    const response = await axios.post(
        "http://localhost:8080/api/auth/login",
        {
          email: email.trim(),
          password: password.trim(),
          role
        }
    );
    return response.data;
  } catch (error) {
    throw error.response?.data || error.message || 'Password is not matched';
  }
};

/**
 * Save auth data to local storage
 * @param {Object} authData - Auth data with tokens and user info
 */
export const saveAuthData = (authData) => {
  localStorage.setItem('token', authData.access_token);
  localStorage.setItem('refreshToken', authData.refresh_token);
  localStorage.setItem('user', JSON.stringify(authData.user));
};

/**
 * Get auth data from local storage
 * @returns {Object} Auth data with token and user info
 */
export const getAuthData = () => {
  const token = localStorage.getItem('token');
  const refreshToken = localStorage.getItem('refreshToken');
  const user = JSON.parse(localStorage.getItem('user') || '{}');
  
  return { token, refreshToken, user };
};

/**
 * Clear auth data from local storage (logout)
 */
export const logout = () => {
  localStorage.removeItem('token');
  localStorage.removeItem('refreshToken');
  localStorage.removeItem('user');
}; 

File: IOS/Features/Auth/AuthService.swift
import Foundation

/// Handles authentication related network operations.
final class AuthService {
    private let api = APIClient.shared
    private let authStorageKey = "authData"

    // MARK: - Networking
    func sendVerificationCode(email: String) async throws {
        let body = try JSONEncoder().encode(["email": email])
        let _: EmptyResponse = try await api.request("api/auth/send-verification-code", method: "POST", body: body)
    }

    func verifyCode(email: String, token: String) async throws -> VerificationResponse {
        let body = try JSONEncoder().encode(["email": email, "token": token])
        let response: VerificationResponse = try await api.request("api/auth/verify-verification-code", method: "POST", body: body)
        return response
    }

    func updateUser(email: String, password: String, role: String) async throws {
        let body = try JSONEncoder().encode(["email": email, "password": password, "role": role])
        let _: EmptyResponse = try await api.request("api/auth/update-user", method: "POST", body: body)
    }

    func login(email: String, password: String, role: String) async throws -> AuthData {
        let body = try JSONEncoder().encode(["email": email, "password": password, "role": role])
        let auth: AuthData = try await api.request("api/auth/login", method: "POST", body: body)
        api.setAuthData(auth)
        return auth
    }

    // MARK: - Auth Storage
    func saveAuthData(_ data: AuthData) {
        if let encoded = try? JSONEncoder().encode(data) {
            UserDefaults.standard.set(encoded, forKey: authStorageKey)
        }
        api.setAuthData(data)
    }

    func getAuthData() -> AuthData? {
        guard let data = UserDefaults.standard.data(forKey: authStorageKey) else { return nil }
        let auth = try? JSONDecoder().decode(AuthData.self, from: data)
        if let auth { api.setAuthData(auth) }
        return auth
    }

    func logout() {
        UserDefaults.standard.removeObject(forKey: authStorageKey)
        api.setAuthData(nil)
    }
}

private struct EmptyResponse: Decodable {}

